File: 15.3-sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()

        ans = []
        i = 0

        while i < len(nums) - 2:

            current = nums[i]
            l , r = i + 1, len(nums) - 1

            while l < r:
                left = nums[l]
                right = nums[r]
                summ = current + left + right

                if summ == 0:
                    ans.append([current, left, right])

                    while l < r and left == nums[l]:
                        l+=1

                    while r > l and right == nums[r]:
                        r-=1

                elif summ > 0:
                    r-=1
                else:
                    l+=1

            while i < len(nums) and nums[i] == current:
                i+=1

        return ans

File: 15.3-sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);

        List<List<Integer>> ans = new ArrayList<>();
        int i = 0;

        while (i < nums.length) {

            int left = i + 1;
            int right = nums.length - 1;
            int currentVal = nums[i];
            while (left < right) {
                int lVal = nums[left];
                int rVal = nums[right];

                int sum = currentVal + lVal + rVal;

                if (sum == 0) {
                    ans.add(List.of(currentVal, lVal, rVal));
                    while (left < right && lVal == nums[left])
                        left++;
                    while (right > left && rVal == nums[right])
                        right--;
                } else if (sum > 0) {
                    right--;
                } else {
                    left++;
                }
            }

            while (i < nums.length && currentVal == nums[i]) {
                i++;
            }

        }

        return ans;

    }
}